### PR TITLE
Avoid duplicate `REQUESTED` selection reason

### PR DESCRIPTION
**Pull request by melix**
### Context

This is similar to the fix in #4301, but this time for direct dependencies
instead of constraints.

### CODE REVIEW

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/DependencyResultSerializerTest.groovy [hunk 1]
```diff
@@ -78,6 +78,7 @@ class DependencyResultSerializerTest extends Specification {
         encoder.flush()
         Map<ModuleComponentSelector, ModuleVersionResolveException> map = new HashMap<>()
         map.put(requested, failure)
+        serializer.reset()
```
Comment (ljacomet) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/DependencyResultSerializerTest.groovy]: That's an unrelated change coming from #4344 it seems, correct?
If yes, it may be better to have it in its own commit, with a reference to the original PR / issue.

Comment (melix) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/DependencyResultSerializerTest.groovy]: The test only fails on this branch. I reckon this should have been done in #4344 , it's a leftover, but clearly only this commit makes this test fail.

Comment (melix) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/DependencyResultSerializerTest.groovy]: (to be clear, I don't like commits which leave the codebase in an unstable state, so if I extract it, it needs either a different PR, a bit overkill, or I need to temporarily disable the test in the first commit and re-enable in the second). I took the fast path :)

Comment (ljacomet) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/DependencyResultSerializerTest.groovy]: Ok, that's interesting. And your approach makes sense then.

**Comment by ljacomet** [comment]
LGTM but would make sure build passes.
